[AArch64][PAC] Combine signing with address materialization

atrosinenko opened the pull request:
In pauthtest ABI, it is common to store a pointer signed with address diversity to a heap-allocated object (namely, storing a signed pointer to VTable as part of new object construction). This patch tries to prevent introducing a signing oracle by combining pointer materialization and its (re)signing into a single pseudo instruction which is not expanded until AsmPrinter, if possible. Furthermore, if the discriminator is computed as `blend(addr, imm)`, these components are detected and passed via separate operands of that pseudo instruction, to prevent substitution of the immediate modifier.

One of the typical patterns is materializing an unsigned pointer with `MOVaddr` pseudo and then signing it with `PAC[ID][AB]` instruction, which can be moved far away from `MOVaddr` by one of the passes in the machine pipeline. As the storage address is not a `Constant` value, one cannot simply emit a `ptrauth` constant in the frontend, which would be selected into `MOVaddrPAC` pseudo.

Another pattern is fetching a pointer to VTable from a signed GOT entry using `LOADgotAUTH` pseudo, authenticating and checking it, and then re-signing after adding an offset.

This commit adds an instruction insertion hook for `PAC` pseudo instruction which detects the above patterns and replaces it either with `MOVaddrPAC` or `LOADgotPAC` instruction.

---

Even though the description of `ISD::PtrAuthGlobalAddress` mentions non-constant address discriminators, this seems to be DAGISel-specific and currently unused, so I post-process `PAC` instruction in `AArch64TargetLowering::EmitInstrWithCustomInserter` not to implement this logic twice: in DAGISel and GlobalISel.

Depends on #146489.

atrosinenko committed:
[AArch64][PAC] Precommit tests on merging MOVaddr/LOADgotAUTH with PAC*

atrosinenko commented:
Here is an example of signing oracle observed in llvm-test-suite (for `MOVaddr -> MOVaddrPAC` case). When I compile `MultiSource/Benchmarks/Prolangs-C++/primes/primes.cpp` source with the following command:
```
/path/to/bin/clang++ -target aarch64-linux-pauthtest -march=v8.3-a --sysroot=... -stdlib=libc++ -O2 ../MultiSource/Benchmarks/Prolangs-C++/primes/primes.cpp -S -o- -mllvm -print-after=finalize-isel
```
I get the following (with CFI directives and most comments omitted for brevity):
<details>
<summary> `main` function (without this patch) </summary>

```
main:
// %bb.0:                               // %entry
	pacibsp
	sub	sp, sp, #96
	stp	x29, x30, [sp, #48]
	stp	x22, x21, [sp, #64]
	stp	x20, x19, [sp, #80]
	add	x29, sp, #48
	add	x8, sp, #24
	adrp	x9, _ZTV7counter+16
	add	x9, x9, :lo12:_ZTV7counter+16
	mov	x10, x8
	movk	x10, #53906, lsl #48
	add	x11, sp, #8
	mov	w22, #34465
	pacda	x9, x10
	adrp	x10, _ZTV5sieve+16
	add	x10, x10, :lo12:_ZTV5sieve+16
	movk	x11, #53906, lsl #48
	adrp	x21, _ZTV6filter+16
	add	x21, x21, :lo12:_ZTV6filter+16
	pacda	x10, x11
	mov	w11, #2
	adrp	x19, .L.str
	add	x19, x19, :lo12:.L.str
	movk	w22, #1, lsl #16
	str	w11, [sp, #40]
	stp	x9, xzr, [sp, #24]
	stp	x10, x8, [sp, #8]
.LBB0_1:                                // %do.body
	ldr	x0, [sp, #16]
	ldr	x16, [x0]
	mov	x17, x0
	movk	x17, #53906, lsl #48
	autda	x16, x17
	mov	x17, x16
	xpacd	x17
	cmp	x16, x17
	b.eq	.Lauth_success_0
	brk	#0xc472
.Lauth_success_0:
	ldr	x8, [x16]
	movk	x16, #35396, lsl #48
	blraa	x8, x16
	mov	w20, w0
	mov	w0, #24
	bl	_Znwm                           // x21 can be tampered with by an attacker during the call
	mov	x8, x21                         // x21 contains _ZTV6filter+16 under normal circumstances
	mov	x9, x0
	movk	x9, #53906, lsl #48
	str	w20, [x0, #16]
	pacda	x8, x9                          // <-- signing oracle
	ldr	x9, [sp, #16]
	str	x0, [sp, #16]
	mov	w1, w20
	stp	x8, x9, [x0]
	mov	x0, x19
	bl	printf
	cmp	w20, w22
	b.lt	.LBB0_1
// %bb.2:                               // %do.end
	mov	w0, #10
	bl	putchar
	mov	w0, wzr
	ldp	x20, x19, [sp, #80]
	ldp	x22, x21, [sp, #64]
	ldp	x29, x30, [sp, #48]
	add	sp, sp, #96
	retab
```

</details>

Under normal circumstances, the value of `_ZTV6filter+16` expression is stored to the callee-saved `x21` register at the beginning of the function and then signed with address diversity multiple times. The problem is that function call is performed between computing the address and signing it, so `x21` can be spilled to memory and tampered with by the attacker.

At the Machine IR level, after `finalize-isel` pass this looks like
```
  %20:gpr64common = MOVaddr target-flags(aarch64-page) @_ZTV6filter + 16, target-flags(aarch64-pageoff, aarch64-nc) @_ZTV6filter + 16
  %21:gpr64 = PACDA %20:gpr64common(tied-def 0), killed %19:gpr64common
```

With this patch applied, it gets rewritten like this (`MOVaddr` is expected to be dead-code-eliminated later, if not used):
```
  %20:gpr64common = MOVaddr target-flags(aarch64-page) @_ZTV6filter + 16, target-flags(aarch64-pageoff, aarch64-nc) @_ZTV6filter + 16
  %31:gpr64noip = COPY %19:gpr64common
  MOVaddrPAC @_ZTV6filter + 16, 2, %31:gpr64noip, 0, implicit-def $x16, implicit-def $x17
  %21:gpr64 = COPY $x16
```

This prevents signing oracle in `main` function:

<details>
<summary> `main` function (with this patch) </summary>

```
main:
// %bb.0:                               // %entry
	pacibsp
	sub	sp, sp, #96
	stp	x29, x30, [sp, #48]
	str	x21, [sp, #64]
	stp	x20, x19, [sp, #80]
	add	x29, sp, #48
	add	x8, sp, #24
	mov	w9, #2
	mov	w21, #34465
	mov	x10, x8
	movk	x10, #53906, lsl #48
	adrp	x19, .L.str
	add	x19, x19, :lo12:.L.str
	adrp	x16, _ZTV7counter
	add	x16, x16, :lo12:_ZTV7counter
	add	x16, x16, #16
	pacda	x16, x10
	str	w9, [sp, #40]
	add	x9, sp, #8
	movk	x9, #53906, lsl #48
	stp	x16, xzr, [sp, #24]
	movk	w21, #1, lsl #16
	adrp	x16, _ZTV5sieve
	add	x16, x16, :lo12:_ZTV5sieve
	add	x16, x16, #16
	pacda	x16, x9
	stp	x16, x8, [sp, #8]
.LBB0_1:                                // %do.body
	ldr	x0, [sp, #16]
	ldr	x16, [x0]
	mov	x17, x0
	movk	x17, #53906, lsl #48
	autda	x16, x17
	mov	x17, x16
	xpacd	x17
	cmp	x16, x17
	b.eq	.Lauth_success_0
	brk	#0xc472
.Lauth_success_0:
	ldr	x8, [x16]
	movk	x16, #35396, lsl #48
	blraa	x8, x16
	mov	w20, w0
	mov	w0, #24
	bl	_Znwm
	ldr	x8, [sp, #16]
	mov	x9, x0
	movk	x9, #53906, lsl #48
	mov	w1, w20
	adrp	x16, _ZTV6filter
	add	x16, x16, :lo12:_ZTV6filter
	add	x16, x16, #16
	pacda	x16, x9                     // <-- the address to sign cannot be spilled to memory
	str	w20, [x0, #16]
	stp	x16, x8, [x0]
	str	x0, [sp, #16]
	mov	x0, x19
	bl	printf
	cmp	w20, w21
	b.lt	.LBB0_1
// %bb.2:                               // %do.end
	mov	w0, #10
	bl	putchar
	mov	w0, wzr
	ldp	x20, x19, [sp, #80]
	ldr	x21, [sp, #64]
	ldp	x29, x30, [sp, #48]
	add	sp, sp, #96
	retab
```

</summary>

ojhunt reviewed (commented):
Semantically, this looks reasonable to me, but I'm not really a back end/isel expert so I think @ahmedbougacha needs to look at this.

One thing I'd like to see is a test case that tries to induce an oracle using the intrinsics in C, something akin to

```c
   void* f1(void *p) {
      void* authed = __ptrauth_auth(p, ....);
      __asm__(""); // may be a sufficient barrier?
      return __ptrauth_sign(authed, ...);
   }
   
   void* f2(void *p) {
      void* authed = __ptrauth_auth(p, ....);
      __asm__(""::"r"(authed)); // theoretically does not spill so no oracle
      return __ptrauth_sign(authed, ...);
   }
   
   void* f3(void *p) {
      void* authed = __ptrauth_auth(p, ....);
      __asm__(""::"m"(authed)); // theoretically does spill so would be an oracle, what should happen?
      return __ptrauth_sign(authed, ...);
   }
```

In an ideal world - out of scope for this change - I think we'd like to error out if we ever see a read from memory into an unauthenticated sign, or a spill and subsequent read of an authenticated value. Doing this robustly probably requires recording that a value was authenticated, then in clang producing a warning or error if a developer does something that forces post-auth store to memory, and for llvm/implicitly authed values increasing the keep-in-register priority for those values, and finally if they still end up being spilled giving them a custom auth schema and signing them on spill.

atrosinenko commented:
> One thing I'd like to see is a test case that tries to induce an oracle using the intrinsics in C

@ojhunt Do you suggest adding a real C source to `clang/test` or just adding LLVM IR emitted for C functions like these to LLVM tests introduced by this PR (and checking both MIR and final asm listing - among other things, this would check that dead instructions left by `EmitInstrWithCustomInserter` are eliminated down the pipeline)?

> In an ideal world - out of scope for this change - I think we'd like to error out if we ever see a read from memory into an unauthenticated sign, or a spill and subsequent read of an authenticated value.

IIUC there are a few cases when the former is valid - such as in a dynamic loader, so an opt-out machinery will be needed (such as maybe some function attribute).

ojhunt commented:
> > One thing I'd like to see is a test case that tries to induce an oracle using the intrinsics in C
> 
> @ojhunt Do you suggest adding a real C source to `clang/test` or just adding LLVM IR emitted for C functions like these to LLVM tests introduced by this PR (and checking both MIR and final asm listing - among other things, this would check that dead instructions left by `EmitInstrWithCustomInserter` are eliminated down the pipeline)?

Yes, I was thinking a `clang/test` test, but maybe IR with intervening asm node would be sufficient? I don't know the IR well enough to know how hard it would be to get tests similar to my above pseudo-c :D

> 
> > In an ideal world - out of scope for this change - I think we'd like to error out if we ever see a read from memory into an unauthenticated sign, or a spill and subsequent read of an authenticated value.
> 
> IIUC there are a few cases when the former is valid - such as in a dynamic loader, so an opt-out machinery will be needed (such as maybe some function attribute).

Yeah, if we were to do something like this for generic loads (not just restore from spill) we'd probably want a new "trust me I know what I'm doing" signing intrinsic that is explicitly permitted to bypass such a guard. Note I'm not suggesting anything like this for this PR, it's very much a future work idea :D

atrosinenko commented:
@ojhunt Updated the test cases - turned out, there was actually an issue with `LOADgotAUTH` not being eliminated.

Considering your test cases, I just realized that they should probably be handled by #130807, not by this patch. As for converting them to LLVM IR, `f1` and `f2` are translated by Clang in a rather straightforward manner and `f3` is a bit more tricky:

```llvm
define dso_local ptr @f1(ptr noundef %p) local_unnamed_addr #0 {
entry:
  %0 = ptrtoint ptr %p to i64
  %1 = tail call i64 @llvm.ptrauth.auth(i64 %0, i32 2, i64 1234)
  tail call void asm sideeffect "", ""() #4, !srcloc !9
  %2 = tail call i64 @llvm.ptrauth.sign(i64 %1, i32 3, i64 42)
  %3 = inttoptr i64 %2 to ptr
  ret ptr %3
}

define dso_local ptr @f2(ptr noundef %p) local_unnamed_addr #0 {
entry:
  %0 = ptrtoint ptr %p to i64
  %1 = tail call i64 @llvm.ptrauth.auth(i64 %0, i32 2, i64 1234)
  %2 = inttoptr i64 %1 to ptr
  tail call void asm sideeffect "", "r"(ptr %2) #4, !srcloc !10
  %3 = tail call i64 @llvm.ptrauth.sign(i64 %1, i32 3, i64 42)
  %4 = inttoptr i64 %3 to ptr
  ret ptr %4
}

define dso_local ptr @f3(ptr noundef %p) local_unnamed_addr #0 {
entry:
  %authed = alloca ptr, align 8
  call void @llvm.lifetime.start.p0(i64 8, ptr nonnull %authed) #4
  %0 = ptrtoint ptr %p to i64
  %1 = tail call i64 @llvm.ptrauth.auth(i64 %0, i32 2, i64 1234)
  %2 = inttoptr i64 %1 to ptr
  store ptr %2, ptr %authed, align 8, !tbaa !11
  call void asm sideeffect "", "*m"(ptr nonnull elementtype(ptr) %authed) #4, !srcloc !15
  %3 = load ptr, ptr %authed, align 8, !tbaa !11
  %4 = ptrtoint ptr %3 to i64
  %5 = call i64 @llvm.ptrauth.sign(i64 %4, i32 3, i64 42)
  %6 = inttoptr i64 %5 to ptr
  call void @llvm.lifetime.end.p0(i64 8, ptr nonnull %authed) #4
  ret ptr %6
}
```
which can be further cleaned up.

As this PR does not handle folding "standalone" auth into sign, the following code is emitted for `f3`:
```
f3:
        sub     sp, sp, #16
        mov     x16, x0
        mov     x17, #1234
        autda   x16, x17
        mov     x17, x16
        xpacd   x17
        cmp     x16, x17
        b.eq    .Lauth_success_2
        brk     #0xc472
.Lauth_success_2:
        add     x8, sp, #8
        str     x16, [sp, #8]
        //APP
        //NO_APP
        ldr     x0, [sp, #8]
        mov     w8, #42
        pacdb   x0, x8
        add     sp, sp, #16
        ret
```
On the branch of #130807 the result is similar, which is quite surprising, moreover `x0` is not reloaded after inline asm, meaning there should be no reason not to merge auth+sign.

On the other hand, if I replace inline asm with function call like this:
```c
  void callee(void *);
  void* f3(void *p) {
     void* authed = __builtin_ptrauth_auth(p, 2, 1234);
     callee(authed);
     return __builtin_ptrauth_sign_unauthenticated(authed, 3, 42);
  }
```
... both patches behave as expected: this one calls a function and signs a potentially overwritten register:
```
// ...
        bl      callee
        mov     w8, #42
        pacdb   x19, x8
// ...
```
and with PR #130807, authentication is re-done after function call:
```
f3:
        pacibsp
        stp     x29, x30, [sp, #-32]!
        str     x19, [sp, #16]
        mov     x29, sp
        mov     x16, x0
        mov     x19, x0
        mov     x17, #1234
        autda   x16, x17
        mov     x17, x16
        xpacd   x17
        cmp     x16, x17
        b.eq    .Lauth_success_0
        brk     #0xc472
.Lauth_success_0:
        mov     x0, x16
        bl      callee
        mov     x16, x19
        mov     x17, #1234
        autda   x16, x17
        mov     x17, x16
        xpacd   x17
        cmp     x16, x17
        b.eq    .Lauth_success_1
        brk     #0xc472
.Lauth_success_1:
        mov     x17, #42
        pacdb   x16, x17
        mov     x0, x16
        ldr     x19, [sp, #16]
        ldp     x29, x30, [sp], #32
        retab
```

atrosinenko commented:
Ping.

I have updated this PR to account for possibility of immediate modifier substitution:
```
  movk    x10, #1234, #48
  bl      callee
  adrp    x17, :got:sym
  add     x17, x17, :got_lo12:sym
  pacda   x17, x10
```
In the above example, the discriminator is computed as `blend(addr, 1234)` - while it is impossible to prevent substitution of `addr` in general, it is definitely possible to at least make sure that 1234 is always blended into whatever is stored in `x10` after `callee` returns (in the below code `x10` is used for clarity - with this PR applied `x16` would be used instead):
```
  bl      callee
  adrp    x17, :got:sym
  add     x17, x17, :got_lo12:sym
  movk    x10, #1234, #48
  pacda   x17, x10
```

In their current state, both this PR and #133788 protect both pointer-to-be-signed as well as the immediate modifier from tampering, but #133788 still has to be updated to handle GlobalISel in addition to DAGISel.

Hopefully, having a unified post-processing at the `MachineInstr` level instead of two separate implementations in DAGISel and GlobalISel could improve maintainability - though, it may be my bias towards `MachineFunction` passes vs. instruction selectors. Furthermore, it feels like with custom inserters it should be possible to extend #132857 to completely get rid of DAGIsel- and GlobalISel-specific selectors for AUT and PAC and use the selectors tablegen-erated from declarative patterns.

Another feature of custom inserters seems to be better handling of cross-basic-block analysis. If I understand it right, at finalize-isel stage virtual registers are in SSA form, and it should be safe to ignore basic blocks while following the chains of copy and increment/decrement instructions, as long as only virtual registers are traversed and no PHI instructions occur.

```llvm
@var = global i32 0
@discvar = global i64 0

define i64 @foo(i1 %cond) {
entry:
  %addrdisc = load i64, ptr @discvar
  %disc = call i64 @llvm.ptrauth.blend(i64 %addrdisc, i64 42)
  %var.ptr.i = ptrtoint ptr @var to i64
  br i1 %cond, label %next, label %exit

next:
  call void asm sideeffect "nop", "r"(i64 %disc)
  br label %exit

exit:
  %signed = call i64 @llvm.ptrauth.sign(i64 %var.ptr.i, i32 2, i64 %disc)
  ret i64 %signed
}

!llvm.module.flags = !{!0}
!0 = !{i32 8, !"ptrauth-elf-got", i32 1}
```
here the DAGISel-based patch is unable to provide a `PtrAuthGlobalAddress` node because discriminator computation is in another basic block (note the `bb.2.exit` basic block in the dump after `finalize-isel` pass):

<details>
<summary>DAGISel-based patch</summary>

```
# ./bin/clang -target aarch64-linux-pauthtest -march=armv8.3-a /tmp/repro.ll -S -o- -O2 -mllvm -print-after=codegenprepare,finalize-isel
warning: overriding the module target triple with aarch64-unknown-linux-pauthtest [-Woverride-module]
        .file   "repro.ll"
*** IR Dump After CodeGen Prepare (codegenprepare) ***
define i64 @foo(i1 %cond) local_unnamed_addr {
entry:
  %addrdisc = load i64, ptr @discvar, align 8
  %disc = tail call i64 @llvm.ptrauth.blend(i64 %addrdisc, i64 42)
  br i1 %cond, label %next, label %exit

next:                                             ; preds = %entry
  tail call void asm sideeffect "nop", "r"(i64 %disc) #1
  br label %exit

exit:                                             ; preds = %next, %entry
  %signed = tail call i64 @llvm.ptrauth.sign(i64 ptrtoint (ptr @var to i64), i32 2, i64 %disc)
  ret i64 %signed
}
# *** IR Dump After Finalize ISel and expand pseudo-instructions (finalize-isel) ***:
# Machine code for function foo: IsSSA, TracksLiveness
Function Live Ins: $w0 in %1

bb.0.entry:
  successors: %bb.1(0x40000000), %bb.2(0x40000000); %bb.1(50.00%), %bb.2(50.00%)
  liveins: $w0
  %1:gpr32 = COPY $w0
  %2:gpr64common = LOADgotAUTH target-flags(aarch64-got) @discvar, implicit-def dead $x16, implicit-def dead $x17, implicit-def dead $nzcv
  %3:gpr64 = LDRXui killed %2:gpr64common, 0 :: (dereferenceable load (s64) from @discvar)
  %4:gpr64 = MOVKXi %3:gpr64(tied-def 0), 42, 48
  %0:gpr64sp = COPY %4:gpr64
  TBZW %1:gpr32, 0, %bb.2
  B %bb.1

bb.1.next:
; predecessors: %bb.0
  successors: %bb.2(0x80000000); %bb.2(100.00%)

  %5:gpr64common = COPY %0:gpr64sp
  INLINEASM &nop [sideeffect] [attdialect], $0:[reguse:GPR64common], %5:gpr64common

bb.2.exit:
; predecessors: %bb.0, %bb.1

  %6:gpr64common = LOADgotAUTH target-flags(aarch64-got) @var, implicit-def dead $x16, implicit-def dead $x17, implicit-def dead $nzcv
  %7:gpr64 = PACDA %6:gpr64common(tied-def 0), %0:gpr64sp
  $x0 = COPY %7:gpr64
  RET_ReallyLR implicit $x0

# End machine code for function foo.

        .text
        .globl  foo                             // -- Begin function foo
        .p2align        2
        .type   foo,@function
foo:                                    // @foo
        .cfi_startproc
// %bb.0:                               // %entry
        adrp    x17, :got_auth:discvar
        add     x17, x17, :got_auth_lo12:discvar
        ldr     x16, [x17]
        autda   x16, x17
        mov     x17, x16
        xpacd   x17
        cmp     x16, x17
        b.eq    .Lauth_success_0
        brk     #0xc472
.Lauth_success_0:
        mov     x8, x16
        ldr     x8, [x8]
        movk    x8, #42, lsl #48
        tbz     w0, #0, .LBB0_2
// %bb.1:                               // %next
        //APP
        nop
        //NO_APP
.LBB0_2:                                // %exit
        adrp    x17, :got_auth:var
        add     x17, x17, :got_auth_lo12:var
        ldr     x16, [x17]
        autda   x16, x17
        mov     x17, x16
        xpacd   x17
        cmp     x16, x17
        b.eq    .Lauth_success_1
        brk     #0xc472
.Lauth_success_1:
        mov     x0, x16
        pacda   x0, x8
        ret
.Lfunc_end0:
        .size   foo, .Lfunc_end0-foo
        .cfi_endproc
                                        // -- End function
        .type   var,@object                     // @var
        .bss
        .globl  var
        .p2align        2, 0x0
var:
        .word   0                               // 0x0
        .size   var, 4

        .type   discvar,@object                 // @discvar
        .globl  discvar
        .p2align        3, 0x0
discvar:
        .xword  0                               // 0x0
        .size   discvar, 8

        .section        ".note.GNU-stack","",@progbits
        .addrsig
        .addrsig_sym var
1 warning generated.
```

</details>

<details>
<summary>Custom inserter-based patch</summary>

```
# ./bin/clang -target aarch64-linux-pauthtest -march=armv8.3-a /tmp/repro.ll -S -o- -O2 -mllvm -print-after=codegenprepare,finalize-isel
warning: overriding the module target triple with aarch64-unknown-linux-pauthtest [-Woverride-module]
        .file   "repro.ll"
*** IR Dump After CodeGen Prepare (codegenprepare) ***
define i64 @foo(i1 %cond) local_unnamed_addr {
entry:
  %addrdisc = load i64, ptr @discvar, align 8
  %disc = tail call i64 @llvm.ptrauth.blend(i64 %addrdisc, i64 42)
  br i1 %cond, label %next, label %exit

next:                                             ; preds = %entry
  tail call void asm sideeffect "nop", "r"(i64 %disc) #1
  br label %exit

exit:                                             ; preds = %next, %entry
  %signed = tail call i64 @llvm.ptrauth.sign(i64 ptrtoint (ptr @var to i64), i32 2, i64 %disc)
  ret i64 %signed
}
# *** IR Dump After Finalize ISel and expand pseudo-instructions (finalize-isel) ***:
# Machine code for function foo: IsSSA, TracksLiveness
Function Live Ins: $w0 in %1

bb.0.entry:
  successors: %bb.1(0x40000000), %bb.2(0x40000000); %bb.1(50.00%), %bb.2(50.00%)
  liveins: $w0
  %1:gpr32 = COPY $w0
  %2:gpr64common = LOADgotAUTH target-flags(aarch64-got) @discvar, implicit-def dead $x16, implicit-def dead $x17, implicit-def dead $nzcv
  %3:gpr64 = LDRXui killed %2:gpr64common, 0 :: (dereferenceable load (s64) from @discvar)
  %4:gpr64 = MOVKXi %3:gpr64(tied-def 0), 42, 48
  %0:gpr64sp = COPY %4:gpr64
  TBZW %1:gpr32, 0, %bb.2
  B %bb.1

bb.1.next:
; predecessors: %bb.0
  successors: %bb.2(0x80000000); %bb.2(100.00%)

  %5:gpr64common = COPY %0:gpr64sp
  INLINEASM &nop [sideeffect] [attdialect], $0:[reguse:GPR64common], %5:gpr64common

bb.2.exit:
; predecessors: %bb.0, %bb.1

  %6:gpr64common = LOADgotAUTH target-flags(aarch64-got) @var, implicit-def dead $x16, implicit-def dead $x17, implicit-def dead $nzcv
  %8:gpr64noip = COPY %3:gpr64
  LOADgotPAC target-flags(aarch64-got) @var, 2, %8:gpr64noip, 42, implicit-def $x16, implicit-def $x17, implicit-def $nzcv
  %7:gpr64 = COPY $x16
  $x0 = COPY %7:gpr64
  RET_ReallyLR implicit $x0

# End machine code for function foo.

        .text
        .globl  foo                             // -- Begin function foo
        .p2align        2
        .type   foo,@function
foo:                                    // @foo
        .cfi_startproc
// %bb.0:                               // %entry
        adrp    x17, :got_auth:discvar
        add     x17, x17, :got_auth_lo12:discvar
        ldr     x16, [x17]
        autda   x16, x17
        mov     x17, x16
        xpacd   x17
        cmp     x16, x17
        b.eq    .Lauth_success_0
        brk     #0xc472
.Lauth_success_0:
        mov     x8, x16
        ldr     x8, [x8]
        tbz     w0, #0, .LBB0_2
// %bb.1:                               // %next
        mov     x9, x8
        movk    x9, #42, lsl #48
        //APP
        nop
        //NO_APP
.LBB0_2:                                // %exit
        adrp    x17, :got_auth:var
        add     x17, x17, :got_auth_lo12:var
        ldr     x16, [x17]
        autda   x16, x17
        mov     x17, x16
        xpacd   x17
        cmp     x16, x17
        b.eq    .Lauth_success_1
        brk     #0xc472
.Lauth_success_1:
        mov     x17, x8
        movk    x17, #42, lsl #48
        pacda   x16, x17
        mov     x0, x16
        ret
.Lfunc_end0:
        .size   foo, .Lfunc_end0-foo
        .cfi_endproc
                                        // -- End function
        .type   var,@object                     // @var
        .bss
        .globl  var
        .p2align        2, 0x0
var:
        .word   0                               // 0x0
        .size   var, 4

        .type   discvar,@object                 // @discvar
        .globl  discvar
        .p2align        3, 0x0
discvar:
        .xword  0                               // 0x0
        .size   discvar, 8

        .section        ".note.GNU-stack","",@progbits
        .addrsig
        .addrsig_sym var
1 warning generated.
```

</details>

atrosinenko committed:
[AArch64][PAC] Combine signing with address materialization

In pauthtest ABI, it is common to store a pointer signed with address
diversity to a heap-allocated object (namely, storing a signed pointer
to VTable as part of new object construction). This patch tries to
prevent introducing a signing oracle by combining pointer
materialization and its (re)signing into a single pseudo instruction
which is not expanded until AsmPrinter, if possible.

One of the typical patterns is materializing an unsigned pointer with
`MOVaddr` pseudo and then signing it with `PAC[ID][AB]` instruction,
which can be moved far away from `MOVaddr` by one of the passes in the
machine pipeline. As the storage address is not a `Constant` value, one
cannot simply emit a `ptrauth` constant in the frontend, which would be
selected into `MOVaddrPAC` pseudo.

Another pattern is fetching a pointer to VTable from a signed GOT entry
using `LOADgotAUTH` pseudo, authenticating and checking it, and then
re-signing after adding an offset.

This commit adds an instruction insertion hook for `PAC[ID][AB]` which
detects the above patterns and replaces it either with `MOVaddrPAC` or
`LOADgotPAC` instruction.

atrosinenko commented:
Ping.

I have rebased this patch so that it is now on top of the stack of
* #145379 - fix for MachineLICM which is needed for LOADgotAUTH to be dead-code-eliminated
* #146488 and #146489 - the patches preventing immediate modifier substitution for sign and for auth/resign operations, respectively. Applying them before this patch instead of after it removes switching on the opcodes of `PAC[DI][AB](Z)?` instructions (which would be dropped later anyway): now the key (as per `AArch64PACKey::ID`) is taken from the operand of `AArch64::PAC` pseudo and the case of zero discriminator is not special anymore. Furthermore, this moves the discussion of `fixupBlendComponents` method to #146488, leaving this PR solely for the discussion of merging address materialization with the subsequent sign operation.

Unfortunately, this PR was initially opened from my fork of llvm-project, so it is impossible to formally add it to the Graphite stack, thus I manually re-targeted it from `main` branch to `users/atrosinenko/pauth-imm-modifier-auth-resign`.

Compared to #133788 this patch has slightly better coverage (supposedly due to cross-BB analysis) and is unified for the two ISel variants. This PR is inherently AArch64-specific, but we don't have other users of PtrAuth so far. Furthermore, implementing the same fix for GlobalISel in addition to DAGISel may require implementing something similar to this patch anyway, I guess.

atrosinenko commented:
Ping.
